// react-window/dist/react-window.js
import { jsx as D } from "react/jsx-runtime";
import { useState as E, useLayoutEffect as J, useEffect as K, useMemo as W, useRef as q, useCallback as V, memo as _, useImperativeHandle as N, createElement as ee } from "react";
function ie(e) {
  let t = e;
  for (; t; ) {
    if (t.dir)
      return t.dir === "rtl";
    t = t.parentElement;
  }
  return !1;
}
function le(e, t) {
  const [s, r] = E(t === "rtl");
  return J(() => {
    e && (t || r(ie(e)));
  }, [t, e]), s;
}
const Y = typeof window < "u" ? J : K;
function Q(e) {
  if (e !== void 0)
    switch (typeof e) {
      case "number":
        return e;
      case "string": {
        if (e.endsWith("px"))
          return parseFloat(e);
        break;
      }
    }
}
function ne({
  box: e,
  defaultHeight: t,
  defaultWidth: s,
  disabled: r,
  element: o,
  mode: i,
  style: n
}) {
  const { styleHeight: c, styleWidth: l } = W(
    () => ({
      styleHeight: Q(n?.height),
      styleWidth: Q(n?.width)
    }),
    [n?.height, n?.width]
  ), [f, a] = E({
    height: t,
    width: s
  }), I = r || i === "only-height" && c !== void 0 || i === "only-width" && l !== void 0 || c !== void 0 && l !== void 0;
  return Y(() => {
    if (o === null || I)
      return;
    const v = new ResizeObserver((k) => {
      for (const x of k) {
        const { contentRect: w, target: z } = x;
        o === z && a((g) => g.height === w.height && g.width === w.width ? g : {
          height: w.height,
          width: w.width
        });
      }
    });
    return v.observe(o, { box: e }), () => {
      v?.unobserve(o);
    };
  }, [e, I, o, c, l]), W(
    () => ({
      height: c ?? f.height,
      width: l ?? f.width
    }),
    [f, c, l]
  );
}
function ce(e) {
  const t = q(() => {
    throw new Error("Cannot call an event handler while rendering.");
  });
  return Y(() => {
    t.current = e;
  }, [e]), V((s) => t.current?.(s), [t]);
}
let j = null;
function fe(e = !1) {
  if (j === null || e) {
    const t = document.createElement("div"), s = t.style;
    s.width = "50px", s.height = "50px", s.overflow = "scroll", s.direction = "rtl";
    const r = document.createElement("div"), o = r.style;
    return o.width = "100px", o.height = "100px", t.appendChild(r), document.body.appendChild(t), t.scrollLeft > 0 ? j = "positive-descending" : (t.scrollLeft = 1, t.scrollLeft === 0 ? j = "negative" : j = "positive-ascending"), document.body.removeChild(t), j;
  }
  return j;
}
function A({
  containerElement: e,
  direction: t,
  isRtl: s,
  scrollOffset: r
}) {
  if (t === "horizontal" && s)
    switch (fe()) {
      case "negative":
        return -r;
      case "positive-descending": {
        if (e) {
          const { clientWidth: o, scrollLeft: i, scrollWidth: n } = e;
          return n - o - i;
        }
        break;
      }
    }
  return r;
}
function R(e, t = "Assertion error") {
  if (!e)
    throw console.error(t), Error(t);
}
function te({
  cachedBounds: e,
  itemCount: t,
  itemSize: s
}) {
  if (t === 0)
    return 0;
  if (typeof s == "number")
    return t * s;
  {
    const r = e.get(
      e.size === 0 ? 0 : e.size - 1
    );
    R(r !== void 0, "Unexpected bounds cache miss");
    const o = (r.scrollOffset + r.size) / e.size;
    return t * o;
  }
}
function de({
  align: e,
  cachedBounds: t,
  index: s,
  itemCount: r,
  itemSize: o,
  containerScrollOffset: i,
  containerSize: n
}) {
  const c = te({
    cachedBounds: t,
    itemCount: r,
    itemSize: o
  }), l = t.get(s), f = Math.max(
    0,
    Math.min(c - n, l.scrollOffset)
  ), a = Math.max(
    0,
    l.scrollOffset - n + l.size
  );
  switch (e === "smart" && (i >= a && i <= f ? e = "auto" : e = "center"), e) {
    case "start":
      return f;
    case "end":
      return a;
    case "center":
      return l.scrollOffset <= n / 2 ? 0 : l.scrollOffset + l.size / 2 >= c - n / 2 ? c - n : l.scrollOffset + l.size / 2 - n / 2;
    case "auto":
    default:
      return i >= a && i <= f ? i : i < a ? a : f;
  }
}
function X({
  cachedBounds: e,
  containerScrollOffset: t,
  containerSize: s,
  itemCount: r,
  overscanCount: o
}) {
  const i = r - 1;
  let n = 0, c = -1, l = 0;
  for (; l < i; ) {
    const f = e.get(l);
    if (f.scrollOffset + f.size > t)
      break;
    l++;
  }
  for (n = l, n = Math.max(0, n - o); l < i; ) {
    const f = e.get(l);
    if (f.scrollOffset + f.size >= t + s)
      break;
    l++;
  }
  return c = Math.min(i, l), c = Math.min(r - 1, c + o), n < 0 ? [0, -1] : [n, c];
}
function ue({
  itemCount: e,
  itemProps: t,
  itemSize: s
}) {
  const r = /* @__PURE__ */ new Map();
  return {
    get(o) {
      for (R(o < e, `Invalid index ${o}`); r.size - 1 < o; ) {
        const n = r.size;
        let c;
        switch (typeof s) {
          case "function": {
            c = s(n, t);
            break;
          }
          case "number": {
            c = s;
            break;
          }
        }
        if (n === 0)
          r.set(n, {
            size: c,
            scrollOffset: 0
          });
        else {
          const l = r.get(n - 1);
          R(
            l !== void 0,
            `Unexpected bounds cache miss for index ${o}`
          ), r.set(n, {
            scrollOffset: l.scrollOffset + l.size,
            size: c
          });
        }
      }
      const i = r.get(o);
      return R(
        i !== void 0,
        `Unexpected bounds cache miss for index ${o}`
      ), i;
    },
    set(o, i) {
      r.set(o, i);
    },
    get size() {
      return r.size;
    }
  };
}
function he({
  itemCount: e,
  itemProps: t,
  itemSize: s
}) {
  return W(
    () => ue({
      itemCount: e,
      itemProps: t,
      itemSize: s
    }),
    [e, t, s]
  );
}
function ae({
  containerSize: e,
  itemSize: t
}) {
  let s;
  switch (typeof t) {
    case "string": {
      R(
        t.endsWith("%"),
        `Invalid item size: "${t}"; string values must be percentages (e.g. "100%")`
      ), R(
        e !== void 0,
        "Container size must be defined if a percentage item size is specified"
      ), s = e * parseInt(t) / 100;
      break;
    }
    default: {
      s = t;
      break;
    }
  }
  return s;
}
function C({
  containerElement: e,
  containerStyle: t,
  defaultContainerSize: s = 0,
  direction: r,
  isRtl: o = !1,
  itemCount: i,
  itemProps: n,
  itemSize: c,
  onResize: l,
  overscanCount: f
}) {
  const [a, I] = E([0, -1]), [v, k] = [
    Math.min(i - 1, a[0]),
    Math.min(i - 1, a[1])
  ], { height: x = s, width: w = s } = ne({
    defaultHeight: r === "vertical" ? s : void 0,
    defaultWidth: r === "horizontal" ? s : void 0,
    element: e,
    mode: r === "vertical" ? "only-height" : "only-width",
    style: t
  }), z = q({
    height: 0,
    width: 0
  }), g = r === "vertical" ? x : w, u = ae({ containerSize: g, itemSize: c });
  J(() => {
    if (typeof l == "function") {
      const d = z.current;
      (d.height !== x || d.width !== w) && (l({ height: x, width: w }, { ...d }), d.height = x, d.width = w);
    }
  }, [x, l, w]);
  const h = he({
    itemCount: i,
    itemProps: n,
    itemSize: u
  }), y = V(
    (d) => h.get(d),
    [h]
  ), B = V(
    () => te({
      cachedBounds: h,
      itemCount: i,
      itemSize: u
    }),
    [h, i, u]
  ), T = V(
    (d) => {
      const m = A({
        containerElement: e,
        direction: r,
        isRtl: o,
        scrollOffset: d
      });
      return X({
        cachedBounds: h,
        containerScrollOffset: m,
        containerSize: g,
        itemCount: i,
        overscanCount: f
      });
    },
    [
      h,
      e,
      g,
      r,
      o,
      i,
      f
    ]
  );
  Y(() => {
    const d = (r === "vertical" ? e?.scrollTop : e?.scrollLeft) ?? 0;
    I(T(d));
  }, [e, r, T]), Y(() => {
    if (!e)
      return;
    const d = () => {
      I((m) => {
        const { scrollLeft: H, scrollTop: G } = e, b = A({
          containerElement: e,
          direction: r,
          isRtl: o,
          scrollOffset: r === "vertical" ? G : H
        }), O = X({
          cachedBounds: h,
          containerScrollOffset: b,
          containerSize: g,
          itemCount: i,
          overscanCount: f
        });
        return O[0] === m[0] && O[1] === m[1] ? m : O;
      });
    };
    return e.addEventListener("scroll", d), () => {
      e.removeEventListener("scroll", d);
    };
  }, [
    h,
    e,
    g,
    r,
    i,
    f
  ]);
  const p = ce(
    ({
      align: d = "auto",
      behavior: m = "auto",
      containerScrollOffset: H,
      index: G
    }) => {
      let b = de({
        align: d,
        cachedBounds: h,
        containerScrollOffset: H,
        containerSize: g,
        index: G,
        itemCount: i,
        itemSize: u
      });
      if (e)
        if (b = A({
          containerElement: e,
          direction: r,
          isRtl: o,
          scrollOffset: b
        }), typeof e.scrollTo == "function")
          r === "horizontal" ? e.scrollTo({
            left: b,
            behavior: m || void 0
          }) : e.scrollTo({
            behavior: m || void 0,
            top: b
          });
        else {
          const O = T(b);
          (O[0] !== v || O[1] !== k) && I(O);
        }
    }
  );
  return {
    getCellBounds: y,
    getEstimatedSize: B,
    scrollToIndex: p,
    startIndex: v,
    stopIndex: k
  };
}
function se(e) {
  return W(() => e, Object.values(e));
}
function Z(e, t) {
  if (e === t)
    return !0;
  if (!!e != !!t || (R(e !== void 0), R(t !== void 0), Object.keys(e).length !== Object.keys(t).length))
    return !1;
  for (const s in e)
    if (!Object.is(t[s], e[s]))
      return !1;
  return !0;
}
function re(e, t) {
  const { style: s, ...r } = e, { style: o, ...i } = t;
  return Z(s, o) && Z(r, i);
}
function ve({
  cellComponent: e,
  cellProps: t,
  className: s,
  columnCount: r,
  columnWidth: o,
  defaultHeight: i = 0,
  defaultWidth: n = 0,
  dir: c,
  gridRef: l,
  onCellsRendered: f,
  onResize: a,
  overscanCount: I = 3,
  rowCount: v,
  rowHeight: k,
  style: x,
  ...w
}) {
  const z = se(t), g = W(
    () => _(e, re),
    [e]
  ), [u, h] = E(null), y = le(u, c), {
    getCellBounds: B,
    getEstimatedSize: T,
    startIndex: p,
    scrollToIndex: d,
    stopIndex: m
  } = C({
    containerElement: u,
    defaultContainerSize: n,
    direction: "horizontal",
    isRtl: y,
    itemCount: r,
    itemProps: z,
    itemSize: o,
    onResize: a,
    overscanCount: I
  }), {
    getCellBounds: H,
    getEstimatedSize: G,
    startIndex: b,
    scrollToIndex: O,
    stopIndex: F
  } = C({
    containerElement: u,
    defaultContainerSize: i,
    direction: "vertical",
    itemCount: v,
    itemProps: z,
    itemSize: k,
    onResize: a,
    overscanCount: I
  });
  N(
    l,
    () => ({
      get element() {
        return u;
      },
      scrollToCell({
        behavior: L = "auto",
        columnAlign: S = "auto",
        columnIndex: M,
        rowAlign: $ = "auto",
        rowIndex: U
      }) {
        O({
          align: $,
          behavior: L,
          containerScrollOffset: u?.scrollTop ?? 0,
          index: U
        }), d({
          align: S,
          behavior: L,
          containerScrollOffset: u?.scrollLeft ?? 0,
          index: M
        });
      },
      scrollToColumn({
        align: L = "auto",
        behavior: S = "auto",
        index: M
      }) {
        d({
          align: L,
          behavior: S,
          containerScrollOffset: u?.scrollLeft ?? 0,
          index: M
        });
      },
      scrollToRow({
        align: L = "auto",
        behavior: S = "auto",
        index: M
      }) {
        O({
          align: L,
          behavior: S,
          containerScrollOffset: u?.scrollTop ?? 0,
          index: M
        });
      }
    }),
    [u, d, O]
  ), K(() => {
    p >= 0 && m >= 0 && b >= 0 && F >= 0 && f && f({
      columnStartIndex: p,
      columnStopIndex: m,
      rowStartIndex: b,
      rowStopIndex: F
    });
  }, [
    f,
    p,
    m,
    b,
    F
  ]);
  const oe = W(() => {
    const L = [];
    if (r > 0 && v > 0)
      for (let S = b; S <= F; S++) {
        const M = H(S);
        for (let $ = p; $ <= m; $++) {
          const U = B($);
          L.push(
            /* @__PURE__ */ ee(
              g,
              {
                ...z,
                columnIndex: $,
                key: `${S}-${$}`,
                rowIndex: S,
                style: {
                  position: "absolute",
                  left: y ? void 0 : 0,
                  right: y ? 0 : void 0,
                  transform: `translate(${y ? -U.scrollOffset : U.scrollOffset}px, ${M.scrollOffset}px)`,
                  height: v > 1 ? M.size : "100%",
                  width: U.size
                }
              }
            )
          );
        }
      }
    return L;
  }, [
    g,
    z,
    r,
    p,
    m,
    B,
    H,
    y,
    v,
    b,
    F
  ]);
  return /* @__PURE__ */ D(
    "div",
    {
      role: "grid",
      ...w,
      className: s,
      dir: c,
      ref: h,
      style: {
        width: "100%",
        height: "100%",
        ...x,
        maxHeight: "100%",
        maxWidth: "100%",
        flexGrow: 1,
        overflow: "auto"
      },
      children: /* @__PURE__ */ D(
        "div",
        {
          className: s,
          style: {
            position: "relative",
            height: G(),
            width: T()
          },
          children: oe
        }
      )
    }
  );
}
const xe = E, we = q;
function me({
  className: e,
  defaultHeight: t = 0,
  listRef: s,
  onResize: r,
  onRowsRendered: o,
  overscanCount: i = 3,
  rowComponent: n,
  rowCount: c,
  rowHeight: l,
  rowProps: f,
  style: a,
  ...I
}) {
  const v = se(f), k = W(
    () => _(n, re),
    [n]
  ), [x, w] = E(null), {
    getCellBounds: z,
    getEstimatedSize: g,
    scrollToIndex: u,
    startIndex: h,
    stopIndex: y
  } = C({
    containerElement: x,
    defaultContainerSize: t,
    direction: "vertical",
    itemCount: c,
    itemProps: v,
    itemSize: l,
    onResize: r,
    overscanCount: i
  });
  N(
    s,
    () => ({
      get element() {
        return x;
      },
      scrollToRow({
        align: T = "auto",
        behavior: p = "auto",
        index: d
      }) {
        u({
          align: T,
          behavior: p,
          containerScrollOffset: x?.scrollTop ?? 0,
          index: d
        });
      }
    }),
    [x, u]
  ), K(() => {
    h >= 0 && y >= 0 && o && o({
      startIndex: h,
      stopIndex: y
    });
  }, [o, h, y]);
  const B = W(() => {
    const T = [];
    if (c > 0)
      for (let p = h; p <= y; p++) {
        const d = z(p);
        T.push(
          /* @__PURE__ */ ee(
            k,
            {
              ...v,
              key: p,
              index: p,
              style: {
                position: "absolute",
                left: 0,
                transform: `translateY(${d.scrollOffset}px)`,
                height: d.size,
                width: "100%"
              }
            }
          )
        );
      }
    return T;
  }, [k, z, c, v, h, y]);
  return /* @__PURE__ */ D(
    "div",
    {
      role: "list",
      ...I,
      className: e,
      ref: w,
      style: {
        ...a,
        maxHeight: "100%",
        flexGrow: 1,
        overflowY: "auto"
      },
      children: /* @__PURE__ */ D(
        "div",
        {
          className: e,
          style: {
            height: g(),
            position: "relative",
            width: "100%"
          },
          children: B
        }
      )
    }
  );
}
const ye = E, be = q;
let P = -1;
function Ie(e = !1) {
  if (P === -1 || e) {
    const t = document.createElement("div"), s = t.style;
    s.width = "50px", s.height = "50px", s.overflow = "scroll", document.body.appendChild(t), P = t.offsetWidth - t.clientWidth, document.body.removeChild(t);
  }
  return P;
}
export {
  ve as Grid,
  me as List,
  Ie as getScrollbarSize,
  xe as useGridCallbackRef,
  we as useGridRef,
  ye as useListCallbackRef,
  be as useListRef
};
